nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/TypeRegistry.java

```diff
@@ -242,7 +243,42 @@ public boolean hasType(String id) {
     }
 
     public Type getType(String id) {
-        return types.get(id);
+        Type type = types.get(id);
+        if (type == null) {
+            return null;
+        } else {
+            Map<String, SubType> allowedSubTypes = getCoreAllowedSubtypes(type);
+            Type newType = type.clone();
+            newType.setAllowedSubTypes(allowedSubTypes);
+            //  do not need to add denied subtypes because allowed subtypes already come filtered from core
+            newType.setDeniedSubTypes(new String[0]);
+            return newType;
+        }
+    }
+
+    /**
+     * @since 8.10
+     */
+    protected Map<String, SubType> getCoreAllowedSubtypes(Type type) {
+        SchemaManagerImpl schemaManager = (SchemaManagerImpl) Framework.getService(SchemaManager.class);
+        Collection<String> coreAllowedSubtypes = schemaManager.getAllowedSubTypes(type.getId());
+        Map<String, SubType> ecmSubTypes = type.getAllowedSubTypes();
+        Map<String, SubType> allowedSubTypes = new HashMap<>();
+
+        SubType subtype;
+        if (coreAllowedSubtypes != null) {
```

Comment:
Use early exit:

    if (coreAllowedSubtypes == null) {
        // there are no subtypes to take care of
        return Collections.emptyMap();
    }

Then the rest unindented. Also move the `allowedSubTypes` and `subtype` declarations closer to their actual use.